sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/implementation/util/StorageBlockingSink.java

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.storage.blob.implementation.util;
+
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Sinks;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Queue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This is a lightweight wrapper around Sinks.Many to implement a Sink backed by a LinkedBlockingQueue, effectively
+ * implementing a BlockingSink.
+ */
+public class StorageBlockingSink {
+    private final ClientLogger logger = new ClientLogger(StorageBlockingSink.class);
+
+    /*
+    Note: Though this is used only by BlockBlobOutputStream, the decision was made to abstract this type out to make
+    testing easier.
+    */
+    private final Sinks.Many<ByteBuffer> writeSink;
+    private static final int WRITE_RETRY_IN_SECONDS = 3;
+
+    /**
+     * Create a new StorageBlockingSink.
+     */
+    public StorageBlockingSink() {
+        /*
+        We use a LinkedBlockingQueue to effectively block the thread that calls tryEmitNext if there is
+        backpressure from downstream. Its capacity is 1 to keep the buffer as small as possible, as downstream
+        implementations do their own buffering.
+        */
+        Queue<ByteBuffer> writeLimitQueue = new LinkedBlockingQueue<>(1);
```

Comment:
shouldn't this be configurable? even if we choose to use 1 I think this should be coming from component that's using it.
As for 1. I'm not sure it's optimal. I'd rather set it to concurrency level of the upload. otherwise we might be hitting internal locks in the queue more often than downstream consumes data contributing to contention. However, to resolve this we'd need to run an experiment like this: pick VM with N cores, upload large file (i.e. 20 GB) with concurrency of N, repeat the process many times for fixed queue size between 1 and N to measure avg and variance. If we could get such data we'd know the answer.